Mathlib/Tactic/TacticAnalysis/Declarations.lean

```diff
@@ -7,83 +7,156 @@ import Mathlib.Tactic.TacticAnalysis
+
+`newTactic stx goal` selects the new tactic to try, which may depend on the old tactic invocation
+in `stx` and the current `goal`.
 -/
-def grindReplacementWith (tacticName : String) (tacticKind : SyntaxNodeKind)
+def terminalReplacement (oldTacticName newTacticName : String) (oldTacticKind : SyntaxNodeKind)
+    (newTactic : Syntax → MVarId → MetaM (TSyntax `tactic))
     (reportFailure : Bool := true) (reportSuccess : Bool := false)
     (reportSlowdown : Bool := false) (maxSlowdown : Float := 1) :
     TacticAnalysis.Config := .ofComplex {
-  out := List MVarId × MessageData
+  out := TerminalReplacementOutcome
   ctx := Syntax
-  trigger _ stx := if stx.getKind == tacticKind
+  trigger _ stx := if stx.getKind == oldTacticKind
     then .accept stx else .skip
   test stx goal := do
-    let tac ← `(tactic| grind)
+    let tac ← newTactic stx goal
     try
       let (goals, _) ← Lean.Elab.runTactic goal tac
-      return (goals, m!"")
-    catch _e => -- Grind throws an error if it fails.
+      match goals with
+      | [] => return .success tac
+      | _ => do
+        let goalsMessages ← goals.mapM fun g => do
+          let e ← instantiateMVars (← g.getType)
+          pure m!"⊢ {MessageData.ofExpr e}\n"
+        return .remainingGoals tac goalsMessages
+    catch _e =>
       let name ← mkAuxDeclName `extracted
       let ((sig, _, modules), _) ← (Mathlib.Tactic.ExtractGoal.goalSignature name goal).run
       let imports := modules.toList.map (s!"import {·}")
-      return ([goal], m!"{"\n".intercalate imports}\n\ntheorem {sig} := by\n  fail_if_success grind\n  {stx}")
+      return .error tac m!"{"\n".intercalate imports}\n\ntheorem {sig} := by\n  fail_if_success {tac}\n  {stx}"
   tell stx _ oldHeartbeats new newHeartbeats :=
-    if let (_ :: _, counterexample) := new then
+    match new with
+    | .error _ msg =>
       if reportFailure then
-        m!"`grind` failed where `{tacticName}` succeeded.\n" ++
-        m!"Original tactic:{indentD stx}\n" ++
-        m!"Counterexample:{indentD counterexample}"
+        let msg :=
+          m!"`{newTacticName}` failed where `{oldTacticName}` succeeded.\n" ++
+          m!"Original tactic:{indentD stx}\n" ++
+          m!"Counterexample:{indentD msg}"
+        return msg
       else
-        none
-    else
+        return none
+    | .remainingGoals newStx goals =>
+      if reportFailure then
+        let msg :=
+          m!"`{newTacticName}` left unsolved goals where `{oldTacticName}` succeeded.\n" ++
+          m!"Original tactic:{indentD stx}\n" ++
+          m!"Replacement tactic:{indentD newStx}\n" ++
+          m!"Unsolved goals:{indentD goals}"
+        return msg
+      else
+        return none
+    | .success newStx => do
+      let msg := if (← liftCoreM <| PrettyPrinter.ppTactic newStx).pretty = newTacticName then
+        m!"`{newTacticName}` can replace `{stx}`"
+      else
+        m!"`{newTacticName}` can replace `{stx}` using `{newStx}`"
```

Comment:
What do you think of formatting it like a "`Try this: {newStx}`" message?

Reply:
Well, it should actually be a Try this, with code action. But I ran out of steam making that happen. I'll add a TODO.